File: leetcode/JavaScript/No126.word-ladder-II.js
/**
 * Difficulty:
 * Hard
 *
 * Desc:
 * Given two words (beginWord and endWord), and a dictionary's word list,
 * find all shortest transformation sequence(s) from beginWord to endWord, such that:
 * 1. Only one letter can be changed at a time
 * 2. Each transformed word must exist in the word list. Note that beginWord is not a transformed word.
 *
 * Example:
 * Given:
 * beginWord = "hit"
 * endWord = "cog"
 * wordList = ["hot","dot","dog","lot","log","cog"]
 * Return
 * [
 *  ["hit","hot","dot","dog","cog"],
 *  ["hit","hot","lot","log","cog"]
 * ]
 *
 * Note:
 * 1. Return an empty list if there is no such transformation sequence.
 * 2. All words have the same length.
 * 3. All words contain only lowercase alphabetic characters.
 * 4. You may assume no duplicates in the word list.
 * 5. You may assume beginWord and endWord are non-empty and are not the same.
 */

 /**
 * @param {string} beginWord
 * @param {string} endWord
 * @param {string[]} wordList
 * @return {string[][]}
 */

var findLadders = function(beginWord, endWord, wordList) {
  if (!beginWord || !endWord || !wordList.length) return [];
  const set = new Set(wordList);

  const res = [];
  const queue = [];
  queue.push([beginWord]);

  while (queue.length && !res.length) {
    for (const list of queue) set.delete(list[list.length - 1]);

    let len = queue.length;
    while (len > 0) {
      const list = queue.shift();
      const word = list[list.length - 1];

      let done = false;
      for (let i = 0; i < word.length && !done; i += 1) {
        for (let j = 97; j <= 122 && !done; j += 1) {
          const letter = String.fromCharCode(j);
          if (letter === word[i]) continue;

          const newWord = `${word.slice(0, i)}${letter}${word.slice(i + 1)}`;
          if (!set.has(newWord)) continue;

          if (newWord === endWord) {
            list.push(newWord);
            res.push(list);
            done = true;
            break;
          }

          queue.push([
            ...list,
            newWord
          ]);
        }
      }
      len -= 1;
    }
  }

  return res;
};

File: leetcode/C#/No126.word-ladder-II.cs
/*
 * Difficulty:
 * Hard
 *
 * Desc:
 * Given two words (beginWord and endWord), and a dictionary's word list,
 * find all shortest transformation sequence(s) from beginWord to endWord, such that:
 * Only one letter can be changed at a time
 * Each transformed word must exist in the word list. Note that beginWord is not a transformed word.
 *
 * Note:
 * Return an empty list if there is no such transformation sequence.
 * All words have the same length.
 * All words contain only lowercase alphabetic characters.
 * You may assume no duplicates in the word list.
 * You may assume beginWord and endWord are non-empty and are not the same.
 *
 * Example 1:
 * Input:
 * beginWord = "hit",
 * endWord = "cog",
 * wordList = ["hot","dot","dog","lot","log","cog"]
 * Output:
 * [
 *  ["hit","hot","dot","dog","cog"],
 *  ["hit","hot","lot","log","cog"]
 * ]
 *
 * Example 2:
 * Input:
 * beginWord = "hit"
 * endWord = "cog"
 * wordList = ["hot","dot","dog","lot","log"]
 * Output: []
 *
 * Explanation: The endWord "cog" is not in wordList, therefore no possible transformation.
*/

public class Solution {

    // timeout
    public IList<IList<string>> FindLadders(string beginWord, string endWord, IList<string> wordList) {
        HashSet<string> set = new HashSet<string>(wordList);
        List<IList<string>> res = new List<IList<string>>();
        if (!set.Contains(endWord)) return res;
        
        Queue<List<string>> q = new Queue<List<string>>();
        q.Enqueue(new List<string>(new string[]{ beginWord }));

        while (q.Count > 0 && res.Count == 0) {

            foreach (List<string> list in q) set.Remove(list[^1]);

            int len = q.Count;
            while (len > 0) {
                List<string> list = q.Dequeue();
                string lastWord = list[^1];

                bool done = false;
                for (int i = 0; i < lastWord.Length && !done; i += 1) {
                    for (int j = 97; j <= 122; j += 1) {
                        if (j == lastWord[i]) continue;

                        string newWord = lastWord.Substring(0, i) + (char)j + lastWord.Substring(i + 1, lastWord.Length - 1 - i);
                        if (!set.Contains(newWord)) continue;

                        if (newWord == endWord) {
                            list.Add(newWord);
                            res.Add(list);
                            done = true;
                            break;
                        }

                        List<string> newList = new List<string>(list);
                        newList.Add(newWord);
                        q.Enqueue(newList);
                    }
                }

                len -= 1;
            }
        }

        return res;
    }
}